### PR TITLE
Rewrite `FindResourceGroupForEnvironment` tests to use a Mock CLI

### PR DESCRIPTION
**Pull request by ellismg**
We had a few end to end tests that only cared about setting up interesting states to test `FindResourceGroupForEnvironment`. Since we now have a good way to mock `AzCli`, we can make these tests much more lightweight and no longer need to worry about invoking `azd` to provision a bunch of infrastructure.

### CODE REVIEW

File: cli/azd/pkg/tools/azcli/mock_azcli.go [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+	"encoding/json"
+	"io"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/containerregistry/armcontainerregistry"
+)
+
+var _ AzCli = &Mock{}
+
+// Mock can be used to build a mock AzCli implementation based on provided functions
+type Mock struct {
```
Comment (weikanglim) [inline cli/azd/pkg/tools/azcli/mock_azcli.go:17]: Curious... Is this approach not taking advantage of `golang`'s duck typing? What if we simply defined the minimal interface that only contained what `azure_resource_manager` depends on `azcli`?

Comment (weikanglim) [reply cli/azd/pkg/tools/azcli/mock_azcli.go:17]: Matt and I chatted offline -- we should create the smaller interfaces naturally as part of the `azcli` refactor down the road.

**Review by weikanglim — APPROVED** [review]

File: cli/azd/pkg/infra/azure_resource_manager_test.go [hunk 2]
```diff
@@ -245,3 +248,96 @@ func TestGetDeploymentResourceOperationsWithNestedDeployments(t *testing.T) {
...
+	}
+
+	tests := []struct {
+		name              string
+		rgsFromFilter     []string
+		rgsFromNoFilter   []string
+		expectedName      string
+		expectedErrorText string
+	}{
+		{"oneTaggedResourceGroup", []string{"custom-rg-name"}, nil, "custom-rg-name", ""},
```
Comment (ellismg) [inline cli/azd/pkg/infra/azure_resource_manager_test.go:281]: In #849 @vhvb1989 wrote:

> From the developer perspective, this is so much better than duplicating all test.
But, from a debugger perspective, this becomes an issue. If there's a test fail in the middle of the loop, you need to update the test array temporarily to run the affected test alone.
>
> During CI, you need to re-run all the loop, there's no way to run just one of the tests.
>
> So, I usually prefer the code duplication, just to help making tests unique.

Happy to go back - but surprised we wouldn't be able to use the `-run` flag of `go test` to solve these problems.  It allows running a single subtest.  In practice I expect we always end up running all the tests in the package so either approach would work.  I find this much nicer when scanning the entire file.

For the VS Code debugging experience I get that it's a little worse, but you can always comment out the block here in the initializer with just the case you care about.

**Review by ellismg — COMMENTED** [review]
Sorry, @vhvb1989 - you ended up commenting on a PR that built on top of this, so I'm moving your issue over here so we can chat about it.

File: cli/azd/pkg/tools/azcli/mock_azcli.go (hunk 1 above)
Comment (wbreza) [inline cli/azd/pkg/tools/azcli/mock_azcli.go:17]: I feel like the mock generation generates just a bunch of extra source code that we don't really need.  Right now it is pretty easy to mock out any tool call including azcli using the mock exec command runner... You can setup different predicates as needed without having to run the actual infra tests each time.

As others mentioned I also find it hard to follow/debug with the nested/looped tests.

Comment (ellismg) [reply cli/azd/pkg/tools/azcli/mock_azcli.go:17]: > I feel like the mock generation generates just a bunch of extra source code that we don't really need. Right now it is pretty easy to mock out any tool call including azcli using the mock exec command runner... You can setup different predicates as needed without having to run the actual infra tests each time.

We no longer use the `az` CLI for these control plane operations anymore, due to your recent refactoring, so I think I can't use a mock command runner to do the validation I want.  Do you have a better idea for this?

Comment (ellismg) [reply cli/azd/pkg/tools/azcli/mock_azcli.go:17]: We chatted in person and Wallace pointed out I can mock the pipeline on AzCli and not have to do all this ceremony, which is great.

Comment (ellismg) [reply cli/azd/pkg/tools/azcli/mock_azcli.go]: @wbreza I rewrote this to do mocking at the HttpClient level and remove the new Mock based on functions type I had tried to add.

File: cli/azd/pkg/infra/azure_resource_manager_test.go (hunk 2 above)
Comment (ellismg) [reply cli/azd/pkg/infra/azure_resource_manager_test.go:294]: As part of responding to Wallace's feedback over what component to mock, I ended up having to do a bit of iteration on the tests.  I found myself just using VS Code to comment out scenarios and while a little frustrating, I think on balance I like it more than a bunch of individual test methods that call into a helper.  I feel like these table-driven tests are idiomatic and we are starting to get some more of them in the codebase, so I'm going to keep what I have here for now.  But I'm happy to go change this across the codebase if we think we get real value from having separate entry points.

**Review by wbreza — APPROVED** [review]
Thanks for making that change to the mocks.  This feels cleaner to me.
